### PR TITLE
Add possibility to revert multiple migrations

### PR DESCRIPTION
**Pull request by romamd**

### CODE REVIEW

**Comment by dhruska** [comment]
It would be awesome get this merged!

**Comment by jrthib** [comment]
Any updates on this?

File: package-lock.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "typeorm",
-  "version": "0.2.18",
+  "version": "0.2.19",
```
Comment (imnotjames) [inline package-lock.json:3]: Probably good to undo this change

File: src/migration/MigrationExecutor.ts [hunk 2]
```diff
@@ -354,9 +432,9 @@ export class MigrationExecutor {
             values["timestamp"] = migration.timestamp;
             values["name"] = migration.name;
         }
-        if (this.connection.driver instanceof MongoDriver) {  
+        if (this.connection.driver instanceof MongoDriver) {
```
Comment (imnotjames) [inline src/migration/MigrationExecutor.ts:435]: Can we drop the style changes

File: src/commands/MigrationRevertCommand.ts [hunk 3]
```diff
@@ -55,7 +60,15 @@ export class MigrationRevertCommand implements yargs.CommandModule {
...
                 transaction: args["t"] === "false" ? false : true
             };
-            await connection.undoLastMigration(options);
+            if (args["until"]) {
+                await connection.undoMigrationsUntil({
+                    ...options,
+                    name: String(args["until"]),
+                });
+            } else {
+                await connection.undoLastMigration(options);
```
Comment (imnotjames) [inline src/commands/MigrationRevertCommand.ts:69]: Should we instead use the new until for this and fetch what the last one is

File: src/migration/MigrationExecutor.ts [hunk 4]
```diff
@@ -247,6 +247,84 @@ export class MigrationExecutor {
...
+            this.connection.logger.logSchemaBuild(`No migration with name ${options.name} was found in the database. Nothing to revert!`);
+            return;
+        }
+
+
+        // get all user's migrations in the source code
+        const allMigrations = this.getMigrations();
+
+        const migrationsToRevert = executedMigrations
+            .filter(migration => migration.id! > targetMigration!.id!)
```
Comment (imnotjames) [inline src/migration/MigrationExecutor.ts:273]: Do we confirm the timestamps are always after one another & can make the assumption the ID order is the timestamp order?

Instead of ID this should be timestamp.  The problem is that the ID always goes up but if we revert and then add more in and then run code again, change values, etc, this may not be correct, will it?

File: src/migration/MigrationExecutor.ts [hunk 5]
```diff
@@ -247,6 +247,84 @@ export class MigrationExecutor {
         }
     }
 
+    async undoMigrationsUntil(options: { name: string }): Promise<Migration[] | void> {
```
Comment (imnotjames) [inline src/migration/MigrationExecutor.ts:250]: Can undo last be pointed to use this function?

File: src/migration/MigrationExecutor.ts [hunk 6]
```diff
@@ -247,6 +247,84 @@ export class MigrationExecutor {
...
+
+        // get all user's migrations in the source code
+        const allMigrations = this.getMigrations();
+
+        const migrationsToRevert = executedMigrations
+            .filter(migration => migration.id! > targetMigration!.id!)
+            .map(migration => allMigrations.find(migrationSource => migrationSource.name === migration.name));
+
+        if (migrationsToRevert.some(migration => !migration)) {
+            this.connection.logger.logSchemaBuild(`Some migrations was found in the database but not in the sourcecode!`);
```
Comment (imnotjames) [inline src/migration/MigrationExecutor.ts:277]: This also needs to throw an error

File: test/functional/migrations/revert-command/command.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
+import "reflect-metadata";
+import {createTestingConnections, closeTestingConnections, reloadTestingDatabases} from "../../../utils/test-utils";
+import {Connection} from "../../../../src/connection/Connection";
+import { expect } from "chai";
+
+describe("migrations > revert until command", () => {
+    let connections: Connection[];
+    before(async () => connections = await createTestingConnections({
+        migrations: [__dirname + "/migration/*.js"],
+        enabledDrivers: ["postgres", "mysql"],
```
Comment (imnotjames) [inline test/functional/migrations/revert-command/command.ts:10]: Can this be run for everything except mongo?

**Review by imnotjames — COMMENTED** [review]
Needs rebase & fix conflicts.

**Comment by canikwe** [comment]
Any update on this?

**Comment by moulinraphael** [comment]
A must have! I'm very interested with such a feature

**Comment by imnotjames** [comment]
Stale for over a year.  Closing.
